File: Q3 (Rafael 2023)/main.c
#include <stdio.h>
#include <string.h>

int compareArrays(int a[], int b[], int n) {
  int ii;
  for(ii = 0; ii < n; ii++) {
    if (a[ii] != b[ii]) 
        return 0;
  }
  return 1;
}
// function to copy an array
void copyArray(int copy[], int arr[], int size)
{
  // loop to iterate through array
  for (int i = 0; i < size; i++)
  {
    copy[i] = arr[i];
  }
}

void print_array(int* arr){
    int size = sizeof(arr)/sizeof(int);
    printf("arr size:%d\n",size);
    for(int i = 0; i< size; i++){
        printf("%d ",arr[i]);
    }
    printf("\n");
}



int measure_quantity_from_jars(int* jars,int size_jars, int* state, int quantity_to_find, int** past_states){
    int size_past_states, result;
    
    if(past_states == NULL){
        size_past_states = 0;
    }
    else{
        size_past_states = sizeof(past_states)/(sizeof(int*));
    }
    printf("STATE:\n");
    print_array(state);

    printf("PAST STATES:\n");
    for(int i =0; i< size_past_states; i++){
        print_array(past_states[i]);
    }

    if(size_jars == 0 || state == NULL){
        printf("error occurs\n");
        return 0;
    }

    //Check if we have the quantity in the states:
    for(int i = 0; i < size_jars; i++){
        if(state[i] == quantity_to_find){
            printf("we found the quantity.\n");
            return 1;
        }
    }

    //Iteration call 1: fill each of the empty jars
    for(int i = 0; i < size_jars; i++){
  


        if(state[i] == 0){
            printf("Fill the jar of size: %d.\n", jars[i]);
            //Allocate memory of the new array.
            //Initialize memroy of the new array.
             //Free the memory of the past array.
            //Check if the current state is in the past array.
            int** new_past_states = malloc((size_past_states + 1) * sizeof(int*));

            for (int i = 0; i < size_past_states; i++) {
                if(compareArrays(state,past_states[i],size_jars)){
                    printf("same state was found\n");
                    // print_status(state, past_states);
                    return 0;
                }   
                new_past_states[i] = malloc(size_jars * sizeof(int));
                copyArray(new_past_states[i], past_states[i],size_jars);
                //free(past_states[i]);

            }
            //free(past_states);

            new_past_states[size_past_states] = malloc(size_jars * sizeof(int));
            copyArray(new_past_states[size_past_states], state,size_jars);
            
            int *new_state = malloc(size_jars * sizeof(int));
            copyArray(new_state, state, size_jars);
            new_state[i] = jars[i];
            //free(state);
            result += measure_quantity_from_jars(jars,size_jars,new_state, quantity_to_find,new_past_states);
        }
    }

    // //Iteration call 2: try to empty each of the full jars
    // for(int i = 0; i < size_jars; i++){
    //     if(state[i] != 0){
    //         printf("Fill the jar of size: %d.\n", jars[i]);
    //         int *new_state = malloc(size_jars * sizeof(int));
    //         copyArray(new_state, state, size_jars);
    //         new_state[i] = jars[i];
    //         measure_quantity_from_jars(jars,size_jars,new_state, quantity_to_find,new_past_states);
    //     }
    // }
    
    return result;
}


int main(){
    int jars[2] = {3,5}; //the size of each jars
    int size_jars = sizeof(jars)/sizeof(int); //the number of jars
    // int *states = malloc(size_jars*sizeof(int)); //the quantity in each of the jars
    // for(int i = 0; i < size_jars; i++){
    //     states[i] = 0;
    // }
    int state[2] = {0,0};
    int **past_states = NULL; //all the past states (2D array)
    int x = measure_quantity_from_jars(jars,size_jars,state,4,past_states);
    printf("%d", x);
    return 0;
}

File: Q3 (Rafael 2023)/main.py
def calculateQuantityFromJars(state,quantity,pastStates):
    newPastStates = pastStates.copy()
    newPastStates.append(state)
    # print(state)

    if(state in pastStates):
        return 0
    
    if(quantity in state.values()):
        print("solution found")
        print(newPastStates)
        return 1

    

    for jar_x,val_x in state.items():
        if(val_x == 0): 
            newState = state.copy()
            newState[jar_x] = jar_x  #fill the empty jar
            if(calculateQuantityFromJars(newState,quantity,newPastStates.copy())): 
                return 1

        for jar_y,val_y in state.items():
            if(jar_x != jar_y and val_x !=0 and val_y!= jar_y): # if the jar are different 
                                                                # and the firt jar is not empty
                                                                # and the second jar is not full.
                newState = state.copy()
                val_transfered = min(val_x ,jar_y - val_y)  # Or we transfered all the quantity of the first jar
                                                            # or we transfered the quantity to fill the second jar.
                newState[jar_x] -= val_transfered
                newState[jar_y] += val_transfered
                if(calculateQuantityFromJars(newState,quantity,newPastStates.copy())): 
                    return 1

        if(jar_x != 0):
            newState = state.copy()
            newState[jar_x] = 0  #empty the non-empty jar
            if(calculateQuantityFromJars(newState,quantity,newPastStates.copy())): 
                return 1

    return 0


def main():
    print("main function called")
    state = dict([(3, 0), (5, 0),(7,0),(100,0)])    #(capacity,quantity)
    quantity_to_find = 97
    pastStates = []        
    calculateQuantityFromJars(state,quantity_to_find,pastStates)

if __name__ == "__main__":
    main()